### PR TITLE
Camp Overview: Edit Camper Information

### PR DESCRIPTION
**Pull request by dhmi19**
## Notion ticket link

[Camp Overview: Edit Camper Information](https://www.notion.so/uwblueprintexecs/Task-Board-F22-9e03bf9f7c4343f09a8a5c16dc48a12a?p=e1dd8cd472ef4ad6a2de3cf1026e01fa&pm=c)

## Implementation description
* Updated the updateCampersByID backend route + function to only update the fields of a camper that can be updated by the client
* Made a folder called `EditCamperModal` which has all the files needed for the edit camper modal functionality. 
* Made a type for editCamperModal which is used by the different files to pass around props 

## Steps to test
1. Go to a camp overview page (http://localhost:3000/admin/camp/63139c7bc3d7b55b44a01531) and click on the hamburger icon next to a registered camper (use mine if you want)
2. Click on the edit button and a modal with all the current info should appear 
3. Try to change all the fields to make sure the results get updated as expected
4. The page should reload after 3 seconds (this might not be the exact functionality we want but the alternative was taking too long to implement for now)

## What should reviewers focus on?
* I removed a lot of code from the backend so just make sure that looks ok 
* Currently, I am refreshing the page after 3 seconds so that the new camper info is loaded after edit. Not sure if this is the best functionality. Ideally we use a global state but I didn't have time to implement that

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [ ] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: frontend/src/components/pages/CampOverview/CampDetails.tsx [hunk 1]
```diff
@@ -88,8 +88,8 @@ const CampDetails = ({ camp, setCamp }: CampDetailsProps): JSX.Element => {
     setCamp({ ...camp, campCounsellors });
   };
 
-  const handleVolunteerChange = (e: any) => {
-    setCamp({ ...camp, volunteers: e.target.value });
+  const handleVolunteerChange = (newVolunteers: string) => {
```
Comment (dhmi19) [inline frontend/src/components/pages/CampOverview/CampDetails.tsx:91]: Not sure why this change got pulled in- it's not part of my changes. Should I remove this from the PR?

File: frontend/src/components/pages/CampOverview/CampDetails.tsx [hunk 2]
```diff
@@ -88,8 +88,8 @@ const CampDetails = ({ camp, setCamp }: CampDetailsProps): JSX.Element => {
...
+    setCamp({ ...camp, volunteers: newVolunteers });
```
Comment (dhmi19) [inline frontend/src/components/pages/CampOverview/CampDetails.tsx:92]: likewise for all the changes in this file

File: frontend/src/components/pages/CampOverview/CampDetails.tsx (hunk 1 above)
Comment (twilkhoo) [reply frontend/src/components/pages/CampOverview/CampDetails.tsx:91]: Hmm its not in dev, is this one of your own changes from another ticket in progress? If so, I guess it'll be more organized to remove it.

File: frontend/src/types/CamperTypes.ts [hunk 3]
```diff
@@ -1,3 +1,11 @@
+export type EmergencyContact = {
```
Comment (twilkhoo) [inline frontend/src/types/CamperTypes.ts:1]: nit: is there a reason why this is "EmergencyContact" and not just "Contact"?

File: backend/typescript/middlewares/validators/camperValidators.ts [hunk 4]
```diff
@@ -211,34 +211,6 @@ export const updateCamperDtoValidator = async (
   if (req.body.allergies && !validatePrimitive(req.body.allergies, "string")) {
     return res.status(400).send(getApiValidationError("allergies", "string"));
   }
-  if (!Array.isArray(req.body.earlyDropoff)) {
```
Comment (twilkhoo) [inline backend/typescript/middlewares/validators/camperValidators.ts:214]: I'm a bit confused can you explain why we're deleting all of this? Is it just not necessary?

File: frontend/src/components/pages/CampOverview/EditCamperModal/index.tsx [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    if (isEditCamperSuccess) {
+      toast({
+        description: `${camper.firstName} ${camper.lastName}'s information has been updated`,
+        status: "success",
+        variant: "subtle",
+        duration: 3000,
+      });
+
+      setTimeout(() => {
+        window.location.reload();
```
Comment (twilkhoo) [inline frontend/src/components/pages/CampOverview/EditCamperModal/index.tsx:95]: I know this is temporary, but if you want a non-reload solution you can check out what I did https://github.com/uwblueprint/focus-on-nature/pull/129
basically I'm triggering a refetch by changing some state in the useEffect's dependency array and I'm sure you could do that here too, it's still not ideal but it's better than an entire page reload (even temporarily).

File: frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperFormResponseSection.tsx [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    question: string,
+    answer: string | undefined,
+    options: string[],
+    questionId: string,
+    isRequired: boolean,
+  ) => {
+    return (
+      <FormControl key={questionId} isRequired={isRequired}>
+        <FormLabel>
+          <b>{question}</b>
```
Comment (twilkhoo) [inline frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperFormResponseSection.tsx:41]: Whereever you use `<b>`, I would reccomend using Chakra's `<Text>` element instead, and passing in some props regarding weight just so we stick to Chakra whereever possible.

File: frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperFormResponseSection.tsx [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+                onChange={() => {
+                  updateFormResponse(question, option);
+                }}
+              >
+                {option}
+              </Radio>
+            ))}
+          </Stack>
+        </RadioGroup>
+        <br />
```
Comment (twilkhoo) [inline frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperFormResponseSection.tsx:58]: Instead of `<br>` add lower padding to the element above (in this case `<RadioGroup>`, same comment whereever else you use `<br>`s to avoid pure HTML elements.

File: frontend/src/APIClients/CamperAPIClient.ts [hunk 8]
```diff
@@ -2,9 +2,38 @@ import { BEARER_TOKEN } from "../constants/AuthConstants";
...
+  camperData: EditCamperInfoFields,
+): Promise<Array<Camper>> => {
+  try {
+    const body = { camperIds: id, ...camperData };
+
+    // Can't send Map<string, string> over HTTP so need to convert to Object before
+    if (camperData.formResponses) {
+      body.formResponses = Object.assign(
+        {},
+        ...Array.from(camperData.formResponses.entries()).map(([k, v]) => ({
```
Comment (twilkhoo) [inline frontend/src/APIClients/CamperAPIClient.ts:21]: nit: Instead of k, v you could use something more descriptive like question, response?

**Review by twilkhoo — COMMENTED** [review]
Most of this looks amazing I left some comments regarding some HTML decisions and the window reloading. One thing I should address though is that testing your branch locally, the modal shows up perfectly and I'm able to edit whatever I need to, but after saving and refreshing the page, the changes don't persist. I'm guessing this works on your end (hence the PR) but just wanted to confirm/doublecheck.

File: frontend/src/types/CamperTypes.ts (hunk 3 above)
Comment (dhmi19) [reply frontend/src/types/CamperTypes.ts:1]: we refer to contacts as Emergency Contacts throughout our product- or at least that's the terminology we used to begin with so I wanted to keep it consistent

File: backend/typescript/middlewares/validators/camperValidators.ts (hunk 4 above)
Comment (dhmi19) [reply backend/typescript/middlewares/validators/camperValidators.ts:214]: these fields can't be edited once the camper is registered. I took them out of the `UpdateCamperDTO` object too. Currently, the validator was looking for those fields but since they can't be edited, it doesn't make sense to include them.

File: frontend/src/components/pages/CampOverview/EditCamperModal/index.tsx (hunk 5 above)
Comment (dhmi19) [reply frontend/src/components/pages/CampOverview/EditCamperModal/index.tsx:95]: That's completely fair but to refresh the state, I would need to change some dependancy in the CampOverview's useEffect so that it can get all the camp data again. I think that could be a little messy and since we're gonna replace with global state anyways, I think it's fine as is for now just to get the ticket moving. 

Good call out though- hadn't thought of that for this use case

File: frontend/src/components/pages/CampOverview/CampDetails.tsx (hunk 1 above)
Comment (dhmi19) [reply frontend/src/components/pages/CampOverview/CampDetails.tsx]: reverted it back to dev's file

**Comment by dhmi19** [comment]
> 

Regarding the update not working: 

Our data is inconsistent in the database again (really need to fix it now since it's a glaring issue affecting everyone). This time, the questions that were under the camper's info did not match the questions that the camp model had (this should never actually happen when we stop playing with the db so often). Hence, the update kept failing. I tested it again with the correct questions setup for the camper and it worked:

https://user-images.githubusercontent.com/57159970/196326019-0b18ec01-392f-4010-a0d7-5fc669f6e9eb.mov

Note: updating the MCQ didn't work on the demo but it's probably because I didn't set up the db correctly. I have tested it before we changed the data for the guelph camp

File: frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperFormResponseSection.tsx (hunk 6 above)
Comment (dhmi19) [reply frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperFormResponseSection.tsx]: done

File: frontend/src/APIClients/CamperAPIClient.ts (hunk 8 above)
Comment (dhmi19) [reply frontend/src/APIClients/CamperAPIClient.ts]: done

File: frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperFormResponseSection.tsx (hunk 7 above)
Comment (dhmi19) [reply frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperFormResponseSection.tsx]: done

File: frontend/src/components/pages/CampOverview/EditCamperModal/index.tsx [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+        <ModalBody overflowY="scroll">
+          <EditCamperModalForm
+            formQuestions={formQuestions}
+            formStateVariables={stateVariables}
+            setStateFuncs={setStateFuncs}
+          />
+        </ModalBody>
+
+        <ModalFooter>
+          <Button onClick={editCamperOnClose} variant="editModalCancel">
```
[the remark this answers is not in the document]
Comment (dhmi19) [reply frontend/src/components/pages/CampOverview/EditCamperModal/index.tsx]: done

**Comment by e-wai** [comment]
@dhmi19 do you mind rebasing your changes? I think there are changes that aren't yours

File: backend/typescript/middlewares/validators/camperValidators.ts (hunk 4 above)
Comment (twilkhoo) [reply backend/typescript/middlewares/validators/camperValidators.ts:214]: Ah now that makes sense to me.

File: frontend/src/components/pages/CampOverview/EditCamperModal/index.tsx (hunk 5 above)
Comment (twilkhoo) [reply frontend/src/components/pages/CampOverview/EditCamperModal/index.tsx:96]: Sounds good.

**Comment by twilkhoo** [comment]
> > 
> 
> Regarding the update not working:
> 
> Our data is inconsistent in the database again (really need to fix it now since it's a glaring issue affecting everyone). This time, the questions that were under the camper's info did not match the questions that the camp model had (this should never actually happen when we stop playing with the db so often). Hence, the update kept failing. I tested it again with the correct questions setup for the camper and it worked:
> 
>  Edit.Camper.Info.WIP.mov 
> Note: updating the MCQ didn't work on the demo but it's probably because I didn't set up the db correctly. I have tested it before we changed the data for the guelph camp

Thanks for the vid, everything else looks good now! I'm giving the LGTM on my end, but be sure to confirm everything with Emma too regarding her comments (rebasing, etc.)

**Review by twilkhoo — APPROVED** [review]
I'm giving the LGTM on my end, but be sure to confirm everything with Emma too regarding her comments (rebasing, etc.), nice job on this!

File: frontend/src/APIClients/CamperAPIClient.ts [hunk 10]
```diff
@@ -2,9 +2,40 @@ import { BEARER_TOKEN } from "../constants/AuthConstants";
...
+
+    // Can't send Map<string, string> over HTTP so need to convert to Object before
+    if (camperData.formResponses) {
+      body.formResponses = Object.assign(
+        {},
+        ...Array.from(camperData.formResponses.entries()).map(
+          ([question, response]) => ({
+            [question]: response,
+          }),
+        ),
```
Comment (e-wai) [inline frontend/src/APIClients/CamperAPIClient.ts:25]: nit: is this different from `new JSONObject(camperData.formResponses);`?

File: frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperFormResponseSection.tsx [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+              : null;
+          case "Text":
+            return renderText(
+              question,
+              camperAnswer,
+              formQuestion.id,
+              formQuestion.required,
+            );
+          default:
+            return null;
```
Comment (e-wai) [inline frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperFormResponseSection.tsx:168]: If this is exhaustive, we should probably do some sort of error here?

File: frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperFormResponseSection.tsx [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+        );
+        const camperAnswer =
+          camperAnswerIndex !== -1
+            ? Object.values(camperResponses)[camperAnswerIndex]
+            : "";
+
+        switch (type) {
+          case "MultipleChoice":
+            return formQuestion.options
+              ? renderMCQ(
```
Comment (e-wai) [inline frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperFormResponseSection.tsx:142]: nit: is ts smart enough that it understands if you put a null check before the switch case (in which it returns null_, and so then you don't have to null check each case?

File: frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperFormResponseSection.tsx [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+    <VStack align="start">
+      {formQuestions.map((formQuestion) => {
+        const { question, type } = formQuestion;
+        const camperAnswerIndex = Object.keys(camperResponses).indexOf(
+          question,
+        );
+        const camperAnswer =
+          camperAnswerIndex !== -1
+            ? Object.values(camperResponses)[camperAnswerIndex]
+            : "";
```
Comment (e-wai) [inline frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperFormResponseSection.tsx:137]: nit: extract to own func

File: frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperModalForm.tsx [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+              </Text>
+            </FormLabel>
+            <Input
+              onChange={(event) => {
+                setStateFuncs.setAge(parseInt(event.target.value, 10));
+              }}
+              value={formStateVariables.age}
+            />
+          </FormControl>
+        </VStack>
```
Comment (e-wai) [inline frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperModalForm.tsx:76]: These are difficult to read, can we create a component for the `VStack` and children, and then pass in appropriate parameters for each response?

File: frontend/src/components/pages/CampOverview/EditCamperModal/index.tsx [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+      toast({
+        description: `${camper.firstName} ${camper.lastName}'s information has been updated`,
+        status: "success",
+        variant: "subtle",
+        duration: 3000,
+      });
+
+      setTimeout(() => {
+        window.location.reload();
+      }, 3000);
```
Comment (e-wai) [inline frontend/src/components/pages/CampOverview/EditCamperModal/index.tsx:97]: Can we move down from 3 secs? Even as the brute force approach - maybe we can't guarantee that we'll cover everyone's network speed, but I suspect this is going to feel like an error for most reasonable users

File: frontend/package.json [hunk 16]
```diff
@@ -24,6 +24,7 @@
     "@types/react": "^17.0.0",
     "@types/react-dom": "^17.0.0",
     "@types/react-jsonschema-form": "^1.7.4",
+    "@types/react-select": "^5.0.1",
```
Comment (e-wai) [inline frontend/package.json:27]: Hmm it's late so I may be missing something, where is this used?

File: frontend/src/components/pages/CampOverview/EditCamperModal/index.tsx [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+    <Modal
+      isOpen={editCamperModalIsOpen}
+      onClose={editCamperOnClose}
+      preserveScrollBarGap
+    >
+      <ModalOverlay />
+      <ModalContent
+        w="1000px"
+        minW="1000px"
+        maxW="1000px"
```
Comment (e-wai) [inline frontend/src/components/pages/CampOverview/EditCamperModal/index.tsx:118]: Do we need the initial width attribute?

File: frontend/src/theme/components/Button.ts [hunk 18]
```diff
@@ -16,6 +16,21 @@ const Button = {
...
+    },
+    editModalSave: {
+      background: "primary.green.100",
+      padding: "12px, 25px, 12px, 25px",
+      radius: "6px",
+      width: "91px",
+      height: "48px",
+      color: "text.white.100",
+    },
   },
```
Comment (e-wai) [inline frontend/src/theme/components/Button.ts:34]: nit: not a huge fan of making these theme components, since we're not reusing them and most of this isn't generally applicable styling?

File: frontend/src/components/pages/CampOverview/EditCamperModal/index.tsx [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+            marginRight="12px"
+          >
+            Cancel
+          </Button>
+          <Button
+            variant="editModalSave"
+            onClick={() => {
+              onSaveClicked();
+            }}
+          >
```
Comment (e-wai) [inline frontend/src/components/pages/CampOverview/EditCamperModal/index.tsx:147]: nit: `onClick={onSaveClicked}`

**Review by e-wai — CHANGES_REQUESTED** [review]
Awesome, awesome changes 🤩 Couple of comments, don't think it's anything structural. The only thing I'll block on is the `react-select` component

Also can you add a screenshot so we expect it to look like?

File: frontend/src/APIClients/CamperAPIClient.ts (hunk 10 above)
Comment (dhmi19) [reply frontend/src/APIClients/CamperAPIClient.ts:25]: I'm not sure which function you are referring to. TS didn't recognize it and I can't seem to find it online. Could you link it for me?

File: frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperFormResponseSection.tsx (hunk 11 above)
Comment (dhmi19) [reply frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperFormResponseSection.tsx:168]: Yeah so this should never really be the case since the value is retrieved from the backend. What kinda error handling were you thinking?

File: frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperFormResponseSection.tsx (hunk 12 above)
Comment (dhmi19) [reply frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperFormResponseSection.tsx:142]: The null check is not in each case, so we would have to break it up for the ones that need options and the TEXT type which doesn't. I can replace the ternary with `&&` to make it a bit cleaner though

File: frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperFormResponseSection.tsx (hunk 13 above)
Comment (dhmi19) [reply frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperFormResponseSection.tsx:137]: done

File: frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperModalForm.tsx (hunk 14 above)
Comment (dhmi19) [reply frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperModalForm.tsx:76]: done

File: frontend/package.json (hunk 16 above)
Comment (dhmi19) [reply frontend/package.json]: I think I got an indication from VS code to update versions at some point. I reverted the file to dev and everything worked, so I made changed it back to dev's version

File: frontend/src/components/pages/CampOverview/EditCamperModal/index.tsx (hunk 17 above)
Comment (dhmi19) [reply frontend/src/components/pages/CampOverview/EditCamperModal/index.tsx:118]: removed

File: frontend/src/theme/components/Button.ts (hunk 18 above)
Comment (dhmi19) [reply frontend/src/theme/components/Button.ts:34]: shfited to inline styling

File: frontend/src/components/pages/CampOverview/EditCamperModal/index.tsx (hunk 19 above)
Comment (dhmi19) [reply frontend/src/components/pages/CampOverview/EditCamperModal/index.tsx:147]: done

File: frontend/src/components/pages/CampOverview/EditCamperModal/index.tsx (hunk 15 above)
Comment (dhmi19) [reply frontend/src/components/pages/CampOverview/EditCamperModal/index.tsx]: changed it to 2 seconds

File: frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperFormResponseSection.tsx (hunk 11 above)
Comment (e-wai) [reply frontend/src/components/pages/CampOverview/EditCamperModal/EditCamperFormResponseSection.tsx:176]: Hmmm I'm not sure what the patterns are here: logging, alert, throw error (don't remember if it's swallowed once deployed?)?

File: frontend/src/components/pages/CampOverview/EditCamperModal/index.tsx (hunk 5 above)
Comment (e-wai) [reply frontend/src/components/pages/CampOverview/EditCamperModal/index.tsx:96]: How about something like an overall `sessionDidChange`? I suggest it in https://github.com/uwblueprint/focus-on-nature/pull/128/files# anyways

https://github.com/uwblueprint/focus-on-nature/blob/29c5d58a7f3726d57a5d823119ae6f488c566602/frontend/src/components/pages/CampOverview/index.tsx#L65-L74

**Review by e-wai — CHANGES_REQUESTED** [review]
There's scroll behaviour when the modal is open, can we avoid that?

Screen Shot 2022-10-24 at 2 17 41 AM

File: frontend/src/APIClients/CamperAPIClient.ts (hunk 10 above)
Comment (dhmi19) [reply frontend/src/APIClients/CamperAPIClient.ts:25]: I tried that with the following code:
```
if (camperData.formResponses instanceof Map) {
      // body.formResponses = Object.assign(
      //   {},
      //   ...Array.from(camperData.formResponses.entries()).map(
      //     ([question, response]) => ({
      //       [question]: response,
      //     }),
      //   ),
      // );

      body.formResponses = Object.fromEntries(camperData.formResponses);
    }
```
But I get the following error:
```
(property) formResponses?: Map<string, string> | undefined
Type '{ [k: string]: string; }' is missing the following properties from type 'Map<string, string>': clear, delete, forEach, get, and 8 more.ts(2740)
```

File: frontend/src/components/pages/CampOverview/EditCamperModal/index.tsx [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+    hasPaid: camper.hasPaid,
+    firstName,
+    lastName,
+    age,
+    allergies,
+    specialNeeds,
+    formResponses,
+  };
+
+  const onSaveClicked = async () => {
```
[the remark this answers is not in the document]
Comment (dhmi19) [reply frontend/src/components/pages/CampOverview/EditCamperModal/index.tsx:87]: addressed delay

**Review by e-wai — APPROVED** [review]
Yeah seems better to me, still think we can go down to 1000 ms or so for refetch delay unless you've had network problems with that?

**Comment by dhmi19** [comment]
> Yeah seems better to me, still think we can go down to 1000 ms or so for refetch delay unless you've had network problems with that?

Which delay are you referring to? I think the refetch call is made immediately after we get a success now no?

**Comment by e-wai** [comment]
@dhmi19 Oh my bad, read wrong -- looks good
